src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs

```diff
@@ -618,445 +501,1122 @@ private static PackOptions GetPackOptions(PackageSpec packageSpec, JObject rawPa
+        {
+            jsonReader.ReadObject(propertyName =>
+            {
+                string propertyValue = jsonReader.ReadAsString();
+
+                packageSpec.PackInclude.Add(new KeyValuePair<string, string>(propertyName, propertyValue));
+            });
         }
 
-        internal static void PopulateCentralDependencies(
-           string packageSpecPath,
-           TargetFrameworkInformation targetFrameworkInformation,
-           JObject properties)
+        private static void ReadPackOptions(PackageSpec packageSpec, JsonTextReader jsonReader, ref bool isMappingsNull)
         {
-            var centralversions = new Dictionary<string, CentralPackageVersion>();
-            var ventralVersionDependenciesProperty = properties["centralPackageVersions"] as JObject;
-            if (ventralVersionDependenciesProperty != null)
+            var wasMappingsRead = false;
+
+            bool isPackOptionsValueAnObject = jsonReader.ReadObject(propertyName =>
             {
-                foreach (var dependency in ventralVersionDependenciesProperty.Values<JToken>())
+                switch (propertyName)
                 {
-                    var depName = ((JProperty)dependency).Name;
-                    var depVersion = (string)((JProperty)dependency).Value;
+                    case "files":
+                        wasMappingsRead = ReadPackOptionsFiles(packageSpec, jsonReader, wasMappingsRead);
+                        break;
 
-                    if (depName == null)
-                    {
-                        throw FileFormatException.Create(
-                            "Unable to resolve central version ''.",
-                            dependency,
-                            packageSpecPath);
-                    }
-                    if (depVersion == null)
-                    {
-                        throw FileFormatException.Create(
-                            "The version cannot be null.",
-                            depName,
-                            packageSpecPath);
-                    }
+                    case "iconUrl":
+                        packageSpec.IconUrl = jsonReader.ReadNextTokenAsString();
+                        break;
+
+                    case "licenseUrl":
+                        packageSpec.LicenseUrl = jsonReader.ReadNextTokenAsString();
+                        break;
 
-                    centralversions.Add(depName, new CentralPackageVersion(depName, VersionRange.Parse(depVersion)));
+                    case "owners":
+                        string[] owners = ReadStringArray(jsonReader);
+
+                        if (owners != null)
+                        {
+                            packageSpec.Owners = owners;
+                        }
+                        break;
+
+                    case "packageType":
+                        ReadPackageTypes(packageSpec, jsonReader);
```

Comment:
ReadPackageTypes() method is different style than all your other Read*() methods. Should your naming pattern differ? (you pass in the packageSpec in this case)

Reply:
Actually, a lot of Read*() methods pass `PackageSpec` instance.  The reason for this is entirely for performance reasons.  `PackageSpec` has many properties which are initialized to an empty collection, and also many properties which are `null` by default.  When the properties are already initialized to a collection, I'd rather use that collection instead of creating _another_ collection and overwriting the `PackageSpec` instance's collection.  The latter practice is what was happening all over the place before this change.

What naming pattern do you suggest?